src/Request/RequestArrayHandler.php

```diff
@@ -19,155 +22,156 @@ public function __construct(
         protected array $array,
         protected ?ParameterErrorCollectingInterface $errors,
         protected string $errorKeyPrefix,
+        protected ?NotSetValue $defaultDefaultValue = null
     ) {
     }
 
-    protected function addError(
-        string $key,
-        ParameterErrorReason $reason,
-    ): void {
-        if (null === $this->errors) {
-            return;
-        }
-
-        $this->errors->addParameterError($this->errorKeyPrefix . $key, $reason);
-    }
-
-    public function getSingleValueFromArray(
+    public function getSingleValue(
         string $key,
         bool $isRequired,
         RequestValueType $type,
         callable $validationFunction = null,
         null|string|float|int|bool|NotSetValue $defaultValue = null,
     ): null|string|float|int|bool|NotSetValue {
-        $exists = array_key_exists($key, $this->array);
-
-        if ($isRequired && !$exists) {
-            $this->addError($key, ParameterErrorReason::MISSING);
-
-            return null;
-        }
-
-        if (!$exists) {
-            return $this->getDefaultValue($type, func_num_args() < 5 ? new NotSetValue() : $defaultValue);
-        }
+        $default = $this->getDefaultValue($type, func_num_args() < 5 ? $this->defaultDefaultValue : $defaultValue);
+        $result  = $default;
 
-        $value = $this->getTypedValue($type, $this->array[$key]);
+        if (!array_key_exists($key, $this->array)) {
+            if ($isRequired) {
+                $this->addError($key, ParameterErrorReason::MISSING);
+            }
+        } else {
+            $result = $this->getTypedValue($type, $this->array[$key]);
 
-        if (null !== $validationFunction && !$validationFunction($value)) {
-            $this->addError($key, ParameterErrorReason::INVALID);
+            if (!empty($validationFunction) && !$validationFunction($result)) {
```

Comment:
Why empty?

Reply:
Easier to read, and more general.
But the most important reason is to get rid of the Yoda Condition. To quote Uncle Bob on this topic:
"Code that protects the author at the expense of the reader is flawed code."

Reply:
To quote Yoda "Do or do not, there is no try". This is not about protecting the author at the expense of the reader it is to protect against a potentially fucking hard to debug bug. And I will keep using Yoda conditions :D.